### PR TITLE
refactorInsecureConnectionConfig

### PR DESCRIPTION
**Pull request by xinlian12**
In previous PR https://github.com/Azure/azure-sdk-for-java/pull/42912, we added a config to allow using insecure connections for emulator. In this PR, we are splitting the config into three different system properties so we can have a more granular config. 

using `COSMOS.HTTP_CONNECTION_WITHOUT_TLS_ALLOWED` or `COSMOS_HTTP_CONNECTION_WITHOUT_TLS_ALLOWED` to allow http connection against emulator. 

Using `COSMOS.EMULATOR_SERVER_CERTIFICATE_VERIFICATION_DISABLED` or `COSMOS_EMULATOR_SERVER_CERTIFICATE_VERIFICATION_DISABLED` to disable cert verification for emulator.

Using `COSMOS.EMULATOR_HOST` or` COSMOS_EMULATOR_HOST` to config the emulator host name. By default it will try to match with `localhost` or `127.0.0.1` or` [::1]` or `[0:0:0:0:0:0:0:1]` .

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java [hunk 1]
```diff
@@ -1327,6 +1333,14 @@ CosmosClientBuilder configs(Configs configs) {
         return this;
     }
 
+    private boolean isEmulatorHost(URI uri) {
+        if (StringUtils.isNotEmpty(Configs.getEmulatorHost())) {
+            return Configs.getEmulatorHost().equals(uri.getHost());
+        }
+
+        return "localhost".equals(uri.getHost()) || "::1".equals(uri.getHost()) || "127.0.0.1".equals(uri.getHost());
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:1341]: "0:0:0:0:0:0:0:1" also valid ipv6 for localhost

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 2]
```diff
@@ -255,27 +256,40 @@ public class Configs {
     private static final boolean DEFAULT_PARTITION_LEVEL_CIRCUIT_BREAKER_DEFAULT_CONFIG_OPT_IN = false;
     private static final String PARTITION_LEVEL_CIRCUIT_BREAKER_DEFAULT_CONFIG_OPT_IN = "COSMOS.PARTITION_LEVEL_CIRCUIT_BREAKER_DEFAULT_CONFIG_OPT_IN";
 
-    // Config to indicate whether allow insecure connections, for example allow http connection or disable cert verification
+    // Config to indicate whether allow http connections for emulator
     // Please note that this config should only during development or test, please do not use in prod env
-    private static final boolean DEFAULT_INSECURE_EMULATOR_CONNECTION_ALLOWED = false;
-    private static final String INSECURE_EMULATOR_CONNECTION_ALLOWED = "COSMOS.INSECURE_EMULATOR_CONNECTION_ALLOWED";
-    private static final String INSECURE_EMULATOR_CONNECTION_ALLOWED_VARIABLE = "COSMOS_INSECURE_EMULATOR_CONNECTION_ALLOWED";
+    private static final boolean DEFAULT_EMULATOR_HTTP_CONNECTION_ALLOWED = false;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:261]: I think "EMULATOR_HTTP_CONNECTION_ALLOWED" should be changed to "HTTP_CONNECTION_WITHOUT_TLS_ALLOWED" ? We don't validate for emulator after all for http?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 3]
```diff
@@ -255,27 +256,40 @@ public class Configs {
...
+
+    // Config to indicate whether disable server certificate validation for emulator
+    // Please note that this config should only during development or test, please do not use in prod env
+    private static final boolean DEFAULT_EMULATOR_CERTIFICATE_VALIDATION_DISABLED = false;
+    private static final String EMULATOR_CERTIFICATE_VALIDATION_DISABLED = "COSMOS.EMULATOR_CERTIFICATE_VALIDATION_DISABLED";
+    private static final String EMULATOR_CERTIFICATE_VALIDATION_DISABLED_VARIABLE = "COSMOS_EMULATOR_CERTIFICATE_VALIDATION_DISABLED";
+
+    // Config to indicate emulator host name
+    // Please note that this config should only during development or test, please do not use in prod env
+    private static final String DEFAULT_EMULATOR_HOST = "localhost";
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:273]: I don't think you need the DEFAULT value here (you use multiple variations of localhost hard-coded as default (which is perefctly fine))?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 4]
```diff
@@ -826,13 +840,31 @@ public static String getCharsetDecoderErrorActionOnUnmappedCharacter() {
...
+
+        return Boolean.parseBoolean(certVerificationDisabledConfig);
+    }
+
+    public static String getEmulatorHost() {
+        return System.getProperty(
+            EMULATOR_HOST,
+            firstNonNull(
+                emptyToNull(System.getenv().get(EMULATOR_HOST_VARIABLE)),
+                DEFAULT_EMULATOR_HOST));
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:868]: either have the ,ocalhost logic being trigegred on localhost as emulatorHost here - or keep using null/empty and return "" here?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:1341]: I think the hostname comparisons shoudl be case-insensitive because the hostname part of Http-uris is always case-insensitive

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 5]
```diff
@@ -15,8 +15,11 @@
...
 * Added system property `COSMOS.ALLOW_UNQUOTED_CONTROL_CHARS` which allow customer to disable/enable `JsonParser.Feature.ALLOW_UNQUOTED_CONTROL_CHARS`. - See [PR 42520](https://github.com/Azure/azure-sdk-for-java/pull/42520)
 * Added system property `COSMOS.CHARSET_DECODER_ERROR_ACTION_ON_MALFORMED_INPUT` and `COSMOS.CHARSET_DECODER_ERROR_ACTION_ON_UNMAPPED_CHARACTER` to allow user config error action on invalid UTF-8 bytes. - See [PR 42520](https://github.com/Azure/azure-sdk-for-java/pull/42520)
-* Added system property `COSMOS.INSECURE_EMULATOR_CONNECTION_ALLOWED` and system variable `COSMOS_INSECURE_EMULATOR_CONNECTION_ALLOWED` to allow using insecure connections to connect to CosmosDB emulator. - See [PR 42912](https://github.com/Azure/azure-sdk-for-java/pull/42912)
+* Added system property `COSMOS.EMULATOR_HTTP_CONNECTION_ALLOWED` and system variable `COSMOS_EMULATOR_HTTP_CONNECTION_ALLOWED` to allow using http connections to connect to CosmosDB emulator. - See [PR 42972](https://github.com/Azure/azure-sdk-for-java/pull/42972)
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:18]: Name - see below.

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/changefeed/pkversion/IncrementalChangeFeedProcessorTest.java [hunk 6]
```diff
@@ -131,6 +131,7 @@ public void readFeedDocumentsStartFromBeginning() throws InterruptedException {
                     .setLeasePrefix("TEST")
                     .setMaxItemCount(10)
                     .setStartFromBeginning(true)
+                    .setStartTime()
                     .setMaxScaleCount(0) // unlimited
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/changefeed/pkversion/IncrementalChangeFeedProcessorTest.java:135]: Why is this needed?

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java]: yea correct - missed the change here. Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:868]: changed to return empty string here

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md]: updated the name

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/changefeed/pkversion/IncrementalChangeFeedProcessorTest.java (hunk 6 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/rx/changefeed/pkversion/IncrementalChangeFeedProcessorTest.java]: good catch - forgot to revert the change as was testing something else

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

**Review by JimSuplizio — APPROVED** [review]
